File: Java/SinglyLinkedList/src/LinkedList.java

public class LinkedList {
    private Node head;
    private Node tail;
    private int size;

    public LinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public void add(int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.size++;
    }

    public void remove(int data) {
        Node current = this.head;
        Node previous = null;
        while (current != null) {
            if (current.data == data) {
                if (previous == null) {
                    this.head = current.next;
                } else {
                    previous.next = current.next;
                }
                this.size--;
                return;
            }
            previous = current;
            current = current.next;
        }
    }

    public void print() {
        Node current = this.head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
}

File: javascript/linkedlist/LinkedList.ts
import Node from "./Node";


class SingllyLinkedList {
    private head: Node | null;
    private tail: Node | null;
    private length: number;

    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    add(data: Number) {
        let newNode = new Node(data, null);
        if (this.head === null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail?.setNext(newNode);
            this.tail = newNode;
        }
        this.length++;
    }

    remove(data: Number) {
        let currentNode = this.head;
        let previousNode: Node | null = null;
        while (currentNode !== null) {
            if (currentNode.getData() === data) {
                if (previousNode !== null) {
                    previousNode.setNext(currentNode.getNext());
                } else {
                    this.head = currentNode.getNext();
                }
                this.length--;
            }
            previousNode = currentNode;
            currentNode = currentNode.getNext();
        }
    }

    reverse() {
        let currentNode = this.head;
        let previousNode: Node | null = null;
        let nextNode: Node | null = null;
        while (currentNode !== null) {
            nextNode = currentNode.getNext();
            currentNode.setNext(previousNode);
            previousNode = currentNode;
            currentNode = nextNode;
        }
        this.tail = this.head;
        this.head = previousNode;
    }
}

export default SingllyLinkedList;